### PR TITLE
Reject unknown options

### PR DESCRIPTION
**Pull request by bjwbell**
Fixes https://github.com/projectcalico/libnetwork-plugin/issues/122

### CODE REVIEW

File: driver/network_driver.go [hunk 1]
```diff
@@ -79,6 +80,25 @@ func (d NetworkDriver) CreateNetwork(request *network.CreateNetworkRequest) erro
...
+	if ok && ipv6 {
+		err := errors.New("Calico driver does not support the --enable_ipv6 flag.")
+		log.Errorln(err)
+		return err
+
+	}
+
+	// Reject all other options
+	for k, v := range request.Options {
+		if k == "com.docker.network.generic" || k == "com.docker.network.internal" || k == "com.docker.network.enable_ipv6" {
```
Comment (tomdee) [inline driver/network_driver.go]: Why are you repeating IPv6 (from above?)

Comment (tomdee) [inline driver/network_driver.go]: Where does the .generic and .internal suffixes come from? Would it be possible to find _all_ unknown options in a more generic way?

**Comment by tomdee** [comment]
Could you add a test which hits this new code?

File: driver/network_driver.go (hunk 1 above)
Comment (bjwbell) [inline driver/network_driver.go]: If it's not then an error is returned when enable_ipv6=false, which is incorrect.

**Comment by tomdee** [comment]
Can you squash your commits - I only want a single commit per PR at the point that I hit merge

File: driver/network_driver.go [hunk 2]
```diff
@@ -61,6 +62,11 @@ func (d NetworkDriver) GetCapabilities() (*network.CapabilitiesResponse, error)
 func (d NetworkDriver) CreateNetwork(request *network.CreateNetworkRequest) error {
 	logutils.JSONMessage("CreateNetwork", request)
 
+	knownOpts := map[string]bool{
+		"com.docker.network.generic":     true,
+		"com.docker.network.internal":    true,
+		"com.docker.network.enable_ipv6": true}
+
 	genericOpts, ok := request.Options["com.docker.network.generic"]
```
Comment (tomdee) [inline driver/network_driver.go]: This overlaps with the new code you added starting on line 98. I don't think you need both bits of code.

**Comment by tomdee** [comment]
The IPv6 stuff doesn't really belong in this PR - ipv6 isn't an unknown option.

**Comment by tomdee** [comment]
Also, there's still no testing of rejecting unknown options. I'm open to not having tests if they're not good value (i.e. expensive to write). But can we be explicit about that?

File: driver/network_driver.go (hunk 2 above)
Comment (bjwbell) [inline driver/network_driver.go]: How's it overlap?

**Comment by bjwbell** [comment]
Removing enable_ipv6 yields https://semaphoreci.com/calico/libnetwork-plugin/branches/pull-request-126/builds/5

**Comment by tomdee** [comment]
OK, so --enable-ipv6 _isn't_ an unknown option because it's always set? Arguably returning an error when it's set to true and the corresponding test for this isn't part of this PR, but since it's done I don't mind it too much being there. 

However, there are still my other comments to resolve.

**Comment by bjwbell** [comment]
Yes, it's always set (in this case to false). 

I'll try getting a good test case for actual unknown options. If there isn't a way pass them in from docker, I'll add a comment with why testing isn't feasible.

File: driver/network_driver.go (hunk 2 above)
Comment (tomdee) [inline driver/network_driver.go]: It returns an error to the user if any com.docker.network.generic options are supplied. Your new code ignores that option and returns a slightly different error when different options are supplied. It's wrong to have two different bits of code doing more or less the same thing.

Comment (bjwbell) [inline driver/network_driver.go]: Removing yields https://semaphoreci.com/calico/libnetwork-plugin/branches/pull-request-126/builds/9

File: driver/network_driver.go [hunk 3]
```diff
@@ -79,6 +74,26 @@ func (d NetworkDriver) CreateNetwork(request *network.CreateNetworkRequest) erro
 		return err
 	}
 
+	// Calico driver does not support ipv6
+	ipv6, ok := request.Options["com.docker.network.enable_ipv6"].(bool)
```
Comment (tomdee) [inline driver/network_driver.go]: It doesn't look like this code is being hit from the test case - the error message is different?

File: tests/default_environment/libnetwork_test.go [hunk 4]
```diff
@@ -63,7 +68,7 @@ var _ = Describe("Libnetwork Tests", func() {
 			It("rejects --opt being used", func() {
 				session := DockerSession("docker network create $RANDOM --opt REJECT -d calico --ipam-driver calico-ipam")
 				Eventually(session).Should(Exit(1))
-				Eventually(session.Err).Should(Say("Error response from daemon: NetworkDriver.CreateNetwork: Arbitrary options are not supported"))
+				Eventually(session.Err).Should(Say("Error response from daemon: NetworkDriver.CreateNetwork: Calico driver does not recognize option com.docker.network.generic (map[REJECT:])"))
```
Comment (tomdee) [inline tests/default_environment/libnetwork_test.go]: The user visible output isn't great here. Are you sure that %v is the best way to represent the unknown "values"

**Comment by tomdee** [comment]
LGTM

**Pull request merged**